File: python/src/schema/insights.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, RootModel
from uuid import UUID
from src.schema.query import InsightQuery


class User(BaseModel):
    id: int
    uuid: UUID
    distinct_id: str
    first_name: str
    email: str


class Insight(BaseModel):
    id: int
    name: str
    description: Optional[str] = None
    filters: Dict[str, Any]
    query: Optional[Dict[str, Any]] = None
    result: Optional[Any] = None
    created_at: str
    updated_at: str
    created_by: Optional[User] = None
    saved: bool
    favorited: Optional[bool] = None
    deleted: bool
    dashboard: Optional[int] = None
    layouts: Optional[Dict[str, Any]] = None
    color: Optional[str] = None
    last_refresh: Optional[str] = None
    refreshing: Optional[bool] = None
    tags: Optional[List[str]] = None


class CreateInsightInput(BaseModel):
    name: str
    query: InsightQuery
    description: Optional[str] = None
    saved: bool = True
    favorited: bool = False
    tags: Optional[List[str]] = None


class UpdateInsightInput(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    filters: Optional[Dict[str, Any]] = None
    query: Optional[InsightQuery] = None
    saved: Optional[bool] = None
    favorited: Optional[bool] = None
    dashboard: Optional[int] = None
    tags: Optional[List[str]] = None


class ListInsights(BaseModel):
    limit: Optional[int] = None
    offset: Optional[int] = None
    saved: Optional[bool] = None
    favorited: Optional[bool] = None
    search: Optional[str] = None


class SQLInsightResponseItem(BaseModel):
    type: str
    data: Dict[str, Any]


class SQLInsightResponse(RootModel[List[SQLInsightResponseItem]]):
    root: List[SQLInsightResponseItem]

File: typescript/src/schema/insights.ts
import { z } from "zod";
import { InsightQuerySchema } from "./query";

export const InsightSchema = z.object({
	id: z.number(),
	short_id: z.string(),
	name: z.string().nullish(),
	description: z.string().nullish(),
	filters: z.record(z.any()),
	query: z.any().nullish(),
	result: z.any().optional(),
	created_at: z.string(),
	updated_at: z.string(),
	created_by: z
		.object({
			id: z.number(),
			uuid: z.string().uuid(),
			distinct_id: z.string(),
			first_name: z.string(),
			email: z.string(),
		})
		.optional()
		.nullable(),
	favorited: z.boolean().nullish(),
	deleted: z.boolean(),
	dashboard: z.number().nullish(),
	layouts: z.record(z.any()).nullish(),
	color: z.string().nullish(),
	last_refresh: z.string().nullish(),
	refreshing: z.boolean().nullish(),
	tags: z.array(z.string()).nullish(),
});

export const SimpleInsightSchema = InsightSchema.pick({
	id: true,
	name: true,
	short_id: true,
	description: true,
	filters: true,
	query: true,
	created_at: true,
	updated_at: true,
	favorited: true,
});

export const CreateInsightInputSchema = z.object({
	name: z.string(),
	query: z.any(),
	description: z.string().optional(),
	favorited: z.boolean().default(false),
	tags: z.array(z.string()).optional(),
});

export const UpdateInsightInputSchema = z.object({
	name: z.string().optional(),
	description: z.string().optional(),
	filters: z.record(z.any()).optional(),
	query: z.record(z.any()).optional(),
	favorited: z.boolean().optional(),
	dashboard: z.number().optional(),
	tags: z.array(z.string()).optional(),
});

export const ListInsightsSchema = z.object({
	limit: z.number().optional(),
	offset: z.number().optional(),
	favorited: z.boolean().optional(),
	search: z.string().optional(),
});

export type PostHogInsight = z.infer<typeof InsightSchema>;
export type CreateInsightInput = z.infer<typeof CreateInsightInputSchema>;
export type UpdateInsightInput = z.infer<typeof UpdateInsightInputSchema>;
export type ListInsightsData = z.infer<typeof ListInsightsSchema>;
export type SimpleInsight = z.infer<typeof SimpleInsightSchema>;

export const SQLInsightResponseSchema = z.array(
	z.object({
		type: z.string(),
		data: z.record(z.any()),
	}),
);

export type SQLInsightResponse = z.infer<typeof SQLInsightResponseSchema>;
